### PR TITLE
mcp23xxx: make devicetree compatibles unique per chip

### LINKED ISSUE #56280: mcp23xxx driver incorrectly assumes all chips in family have push-pull outputs
**Describe the bug**

The `mcp23xxx` driver [only allows push-pull pins](https://github.com/zephyrproject-rtos/zephyr/blob/9c69f29d0d37aa608b575154696aba088fc2a15e/drivers/gpio/gpio_mcp23xxx.c#L182-L185). This is correct for the MCP23008, MCP23016 and MCP23017, but the MCP23009 and MCP23018 have open-drain outputs instead of push-pull outputs.

**Expected behavior**
The driver and device bindings should be adjusted to match the real hardware capabilities.

**Impact**
Currently, open-drain outputs have to be configured as if they were push-pull outputs, which can make the devicetree slightly confusing to read (e.g. `GPIO_ACTIVE_HIGH | GPIO_PULL_UP`, which should be `GPIO_OPEN_DRAIN | GPIO_PULL_UP`).

### PR DESCRIPTION
**Pull request by arbrauns**
Pin definitions should correctly reflect the actual drive mode of the GPIO controller, either push-pull or open drain.

EDIT: see https://github.com/zephyrproject-rtos/zephyr/pull/65797#issuecomment-1993904939 for current description of the changes

Fixes #56280.

### CODE REVIEW

File: dts/bindings/gpio/microchip,mcp23xxx.yaml [hunk 1]
```diff
@@ -31,6 +31,12 @@ properties:
     description: |
       Number of gpios supported by the chip.
 
+  open-drain:
+    type: boolean
+    description: |
+      Specifies that the chip has open-drain rather than push-pull outputs. Should be set
+      for mcp23x09 and mcp23x18 chips.
```
Comment (henrikbrixandersen) [inline dts/bindings/gpio/microchip,mcp23xxx.yaml:38]: Sounds like we ought to have separate bindings for these chips instead?

Comment (arbrauns) [reply dts/bindings/gpio/microchip,mcp23xxx.yaml:38]: Maybe, but I haven't found a nice way to do that without needlessly copying a bunch of code/configuration when the only difference is really just a single bit.

**Comment by arbrauns** [comment]
I've made quite a few changes now:

1. The legacy `mcp23s17` driver has been removed, since its functionality is covered by `mcp23xxx`
2. Devicetree bindings have been changed to exact chip types (e.g. `microchip,mcp23s09` instead of `microchip,mcp23sxx`)
3. The number of GPIOs and the output type (OD/PP) is now automatically inferred from the chip type, removing the need for the `ngpio` devicetree property

This is a breaking change (users of the chips have to change their devicetree compatibles and remove the obsolete properties), but I think it's more convenient overall.

**Comment by arbrauns** [comment]
@jvasanth1 could I get a review?

**Comment by jvasanth1** [comment]
@con-pax could you help to review?

**Comment by con-pax** [comment]
> @jvasanth1 could I get a review?

Hey, apologies for the delay. Will do

**Comment by jvasanth1** [comment]
Hi @MaureenHelm, looks like Microchip is not the code owners for the mcp23xxx gpio controller drivers

**Comment by MaureenHelm** [comment]
> Hi @MaureenHelm, looks like Microchip is not the code owners for the mcp23xxx gpio controller drivers

How so? The zephyrbot assigned this PR to you based on the MAINTAINERS file, so it looks like you're the code owner to me.

**Comment by jvasanth1** [comment]
> > Hi @MaureenHelm, looks like Microchip is not the code owners for the mcp23xxx gpio controller drivers
> 
> How so? The zephyrbot assigned this PR to you based on the MAINTAINERS file, so it looks like you're the code owner to me.

I think the MAINTAINERS file needs to be updated.

**Comment by arbrauns** [comment]
@con-pax ping

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Please add entries to the migration guide for the removed devicetree compatibles and properties.

File: doc/releases/migration-guide-3.7.rst [hunk 2]
```diff
@@ -335,6 +335,14 @@ Flash
 General Purpose I/O (GPIO)
 ==========================
 
+* The devicetree bindings for the Microchip ``mcp23xxx`` series have been split up. Users of
+  ``microchip,mcp230xx`` and ``microchip,mcp23sxx`` should change their devicetree ``compatible``
+  values to the specific chip variant, e.g. :dtcompatible:`microchip,mcp23017`.
+  The ``ngpios`` devicetree property has been removed, since it is implied by the model name.
+  Chip variants with open-drain outputs (``mcp23x09``, ``mcp23x18``) now correctly reflect this in
+  their driver API, users of these devices should ensure they pass appropriate values to
+  :c:func:`gpio_pin_set`.
```
Comment (aescolar) [inline doc/releases/migration-guide-3.7.rst:344]:
```suggestion
  :c:func:`gpio_pin_set` (:github:`65797`).
```

**Comment by decsny** [comment]
The explosion of DT bindings on various PRs because people are making a new file for every single part number and sub part number is really bothering me. We will have to address this soon with a more refined solution than having a gazillion files

**Comment by arbrauns** [comment]
Sure, some form of templating for binding files would be nice. Having more precise compatible names for devices is good though, imo.

**Comment by decsny** [comment]
> Sure, some form of templating for binding files would be nice. Having more precise compatible names for devices is good though, imo.

yes, we should have the specific part number as the compatible string, that is according to the DT spec, the issue is the zephyr infrastructure around this is not scaling to meet that requirement

**Comment by arbrauns** [comment]
Ping, seems like both @jvasanth1 and @con-pax are unavailable.

**Comment by arbrauns** [comment]
@henrikbrixandersen is there any way I can roll for a new assignee?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Looks good, only one minor nit.

**Comment by henrikbrixandersen** [comment]
Reassigning to myself due to lack of activity from @jvasanth1.

@arbrauns Please rebase on `main` and resolve the merge conflict.

File: dts/bindings/gpio/microchip,mcp23s17.yaml [hunk 3]
```diff
@@ -1,26 +1,11 @@
 #
-# Copyright (c) 2020 Geanix ApS
+# Copyright (c) 2024 SILA Embedded Solutions GmbH
```
[the remark this answers is not in the document]
Comment (arbrauns) [reply dts/bindings/gpio/microchip,mcp23s17.yaml:2]: If you check the commits, I completely removed that file and then replaced it with new content that is unrelated to the previous content, therefore the old copyright doesn't apply to it.

**Review by henrikbrixandersen — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Comment by sknsean** [comment]
LGTM 👍

**Review by kartben — APPROVED** [review]

**Pull request merged**